src/cljam/bam/decoder.clj

```diff
@@ -87,61 +92,55 @@
     (= (ref-name refs n) rname) "="
     :else (ref-name refs n)))
 
-(def ^:private cigar-op-map
-  (hash-map (byte 0) \M
-            (byte 1) \I
-            (byte 2) \D
-            (byte 3) \N
-            (byte 4) \S
-            (byte 5) \H
-            (byte 6) \P
-            (byte 7) \=
-            (byte 8) \X))
-
-(defn- decode-cigar-op [op]
-  (get cigar-op-map op))
-
-(defn decode-cigar
+(defn decode-cigar-and-ref-length
   [cigar-bytes]
-  (let [buf (ByteBuffer/wrap cigar-bytes)]
+  (let [buf (ByteBuffer/wrap cigar-bytes)
+        sb (StringBuilder.)]
     (.order buf ByteOrder/LITTLE_ENDIAN)
-    (loop [sb (StringBuilder.)]
+    (loop [ref-length 0]
       (if (.hasRemaining buf)
-        (let [b  (.getInt buf)
+        (let [b (.getInt buf)
               op (bit-and b 0xF)
               n  (bit-shift-right b 4)]
-          (recur (doto sb (.append n)
-                          (.append (decode-cigar-op op)))))
-        (str sb)))))
+          (doto sb
+            (.append n)
+            (.append (case op 0 \M 1 \I 2 \D 3 \N 4 \S 5 \H 6 \P 7 \= 8 \X)))
+          (recur (+ ref-length (case op 0 n 2 n 3 n 7 n 8 n 0))))
+        [(.toString sb) ref-length]))))
+
+(defrecord SAMAlignment [^String qname ^int flag ^String rname ^int pos ^int end ^int mapq ^String cigar
+                         ^String rnext ^int pnext ^int tlen ^String seq ^String qual options])
```

Comment:
I changed `cljam.sam.reader/read-alignments` to return `SAMAlignment`s.
`SAMAlignment` is now defined in `cljam.io` without `^String` type hints.